File: 480.sliding-window-median.java
/*
 * @lc app=leetcode id=480 lang=java
 *
 * [480] Sliding Window Median
 */

// @lc code=start
class Solution {
    public double[] medianSlidingWindow(int[] nums, int k) {
        double[] result = new double[nums.length - k + 1];
        int[] window = new int[k];

        boolean even = k % 2 == 0;
        int mid = k / 2;
        int preMid = mid - 1;

        // get the fisrt k elements form nums[] to window[]
        System.arraycopy(nums, 0, window, 0, k);
        Arrays.sort(window);
        result[0] = even ? ((double) window[preMid] + window[mid]) / 2 : window[mid];

        int startIndex = 0;
        while (startIndex < nums.length - k) {
            // remove nums[startIndex], add nums[startIndex + 1 + k - 1]
            // and then sort
            sortedReplace(window, nums[startIndex], nums[startIndex + 1 + k - 1]);
            startIndex++;
            result[startIndex] = even ? ((double) window[preMid] + window[mid]) / 2 : window[mid];
        }
        return result;
    }

    private void sortedReplace(int[] window, int _old, int _new) {
        // System.out.println(_old + " " + _new);
        // for (int i : window) {
        // System.out.print(i + " ");
        // }
        // System.out.println(" ");

        // remove old
        int idx_old = Arrays.binarySearch(window, _old);
        // move window[idx_old + 1...window.length - 1] backward one unit
        System.arraycopy(window, idx_old + 1, window, idx_old, window.length - idx_old - 1);

        // for (int i : window) {
        // System.out.print(i + " ");
        // }
        // System.out.println(" ");

        // insert next
        int lo = 0, hi = window.length - 2;

        // ????
        // find the first element > _new
        while (lo <= hi) {
            // (left + right) / 2 : 9ms
            // (left + right) >>> 1 : 7ms
            int mid = (lo + hi) >>> 1;
            if (window[mid] < _new) {
                // if window[mid]
                lo = mid + 1;
            } else {
                // why mid - 1 ?
                hi = mid - 1;
            }
        }

        // move window[left...window.length - 2] forward one unit
        System.arraycopy(window, lo, window, lo + 1, window.length - lo - 1);
        window[lo] = _new;

        // System.out.println(lo);
        // for (int i : window) {
        // System.out.print(i + " ");
        // }
        // System.out.println(" ");

    }
    // [1,3,-1,-3,5,3,6,7]\n4
}
// @lc code=end

File: 480.sliding-window-median.py
#
# @lc app=leetcode id=480 lang=python3
#
# [480] Sliding Window Median
#


# @lc code=start
class DualHeap:
    def __init__(self, k):
        self.small, self.large = [], []
        self.small_size, self.large_size = 0, 0
        self.delay = defaultdict(int)
        self.k = k

    def prune(self, heap):
        while heap:
            num = -heap[0] if heap is self.small else heap[0]
            if self.delay[num]:
                self.delay[num] -= 1
                if heap is self.small:
                    heapq.heappop(self.small)
                else:
                    heapq.heappop(self.large)
            else:
                break

    def balance(self):
        # samll can have one more than large
        if self.small_size > self.large_size + 1:
            # move top of small -> large
            val = -heapq.heappop(self.small)
            self.small_size -= 1
            heapq.heappush(self.large, val)
            self.large_size += 1
            self.prune(self.small)
        elif self.small_size < self.large_size:
            val = heapq.heappop(self.large)
            self.large_size -= 1
            heapq.heappush(self.small, -val)
            self.small_size += 1
            self.prune(self.large)

    def insert(self, num):
        if not self.small or num <= -self.small[0]:
            heapq.heappush(self.small, -num)
            self.small_size += 1
        else:
            heapq.heappush(self.large, num)
            self.large_size += 1
        self.balance()

    def erase(self, num):
        # lazy deleete: mark and decrement size counter
        self.delay[num] += 1
        if num <= -self.small[0]:
            self.small_size -= 1
            if num == -self.small[0]:
                self.prune(self.small)
        else:
            self.large_size -= 1
            if self.large and num == self.large[0]:
                self.prune(self.large)
        self.balance()

    def get_median(self):
        if self.k % 2 == 1:
            return -self.small[0]
        else:
            return (-self.small[0] + self.large[0]) / 2


class Solution:
    def medianSlidingWindow(self, nums: List[int], k: int) -> List[float]:
        dh = DualHeap(k)
        res = []
        n = len(nums)
        for i in range(n):
            dh.insert(nums[i])
            if i >= k:
                dh.erase(nums[i - k])
            if i >= k - 1:
                res.append(dh.get_median())

        return res


# @lc code=end
